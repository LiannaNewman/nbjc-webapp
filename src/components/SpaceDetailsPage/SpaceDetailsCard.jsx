import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import cx from 'classnames';
import { useHistory } from 'react-router-dom';

import {
  Button,
  Card,
  CardHeader,
  CardMedia,
  CardContent,
  Divider,
  Typography,
  IconButton,
  Link,
} from '@material-ui/core';
import StarIcon from '@material-ui/icons/Star';
import NavigateNextIcon from '@material-ui/icons/NavigateNext';
import ShareIcon from '@material-ui/icons/Share';

import useMobileDevice from '../../hooks/useMobileDevice';
import ChipList from '../ChipList';

const styles = () => ({
  root: {
    '& .MuiSvgIcon-root': {
      padding: '6px 6px;',
    },
  },
  starIcon: {
    paddingBottom: '0 !important',
  },
  rating: {
    display: 'block',
  },
  headerAction: {
    margin: '0 !important',
  },
  card: {
    justifyContent: 'center',
    alignContent: 'center',
    padding: '15px',
    margin: '30px',
  },
  cardMedia: {
    margin: 'auto',
    height: '300px',
  },
  reviewDiv: {
    width: '100%',
    display: 'inline-flex',
    flexFlow: 'row',
    justifyContent: 'space-between',
  },
  featuredReview: {
    margin: '20px',
    textAlign: 'left',
    display: 'inline-flex',
  },
  seeAllButton: {
    display: 'inline-flex',
    float: 'right',
    margin: 0,
  },
  reviewButton: {
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    margin: '30px',
  },
  // todo: this needs to be a hyperlink
  location: {
    display: 'flex',
    'flex-direction': 'row',
    marginBottom: 15,
    color: 'rgba(0, 0, 0, 0.6)',
  },
  distance: {
    textAlign: 'right',
    flexGrow: 2,
  },
  address: {
    marginRight: 30,
    flexGrow: 1,
    textAlign: 'left',
    color: 'black',
  },
  chipWrapper: {
    'margin-top': '15px',
  },
  filter: {
    margin: '15px 20px 15px 0',
  },
  nextButton: {
    float: 'right',
    padding: 0,
  },
  shareButton: {
    float: 'right',
  },
  subtitles: {
    fontSize: 12,
    textTransform: 'uppercase',
    textAlign: 'left',
    color: 'grey',
    marginTop: '10px',
  },
  mainInformation: {
    color: 'black',
  },
  hoursOfOperation: {
    display: 'flex',
  },
  url: {
    textDeocration: 'none',
  },
});

const SpaceDetailCard = ({
  totalReviews,
  id,
  name,
  category,
  averageRating,
  imageUrl,
  address,
  // distance,
  hoursOfOperation,
  filters,
  phoneNumber,
  website,
  classes,
  yelpUrl,
  overrideClasses,
}) => {
  const [isCopied, setIsCopied] = useState(false);
  const [isMobileOrTablet] = useMobileDevice();

  const history = useHistory();
  /* if desktop, copy to clipboard unique url for the space.
   if mobile, share sheet to text, email, whatever is on your phone or tablet
  */
  const handleShare = () => {
    /* navigator */
    if (isMobileOrTablet && navigator.share) {
      navigator
        .share({
          text: `Check out ${name} at ${document.location.href}`,
        })
        .then(() => {
          console.log('Successfully shared');
        })
        .catch((error) => {
          console.error('Something went wrong sharing the blog', error);
        });
    } else if (navigator.clipboard) {
      setIsCopied(true);
      const currentUrl = document.location.href || window.location.href;
      if (isCopied) {
        navigator.clipboard.writeText(currentUrl);
        console.log('Successfully shared');
      }
    }
  };

  const handleClick = () => {
    const location = {
      pathname: `/spaces/${id}/reviews/`,
      state: { name },
    };
    history.push(location);
  };
  return (
    <Card
      className={cx(classes.root, overrideClasses.root, classes.card)}
      variant="outlined"
      key={id}
    >
      <CardHeader
        avatar={(
          <div color="secondary">
            <StarIcon
              color="secondary"
              fontSize="large"
              classes={{ root: classes.starIcon }}
            />
            <Typography
              variant="caption"
              color="secondary"
              className={classes.rating}
              align="center"
            >
              {averageRating}
            </Typography>
          </div>
        )}
        title={(
          <Typography variant="h6" className={classes.h6}>
            {name}
          </Typography>
        )}
        subheader={<Typography variant="body2">{category}</Typography>}
        classes={{ action: classes.headerAction }}
      />
      <CardMedia
        // component="img"
        image={imageUrl}
        className={classes.cardMedia}
      />
      <CardContent>
        <div className={classes.chipWrapper}>
          <ChipList chips={filters} />
        </div>
        <Divider />
        <Typography variant="h5" className={classes.featuredReview}>
          Recent Reviews
        </Typography>
        {totalReviews > 0 ? (
          <Button
            variant="outlined"
            color="primary"
            onClick={handleClick}
            className={classes.seeAllButton}
          >
            {`See All ${totalReviews}`}
          </Button>
        ) : (
          <Typography variant="body2" align="center">
            There are no reviews. Be the first to rate and review this space!
          </Typography>
        )}
        <div className={classes.reviewButton}>
          <Button
            variant="contained"
            color="primary"
            href={`/spaces/${id}/reviews/new`}
          >
            Write a review
          </Button>
        </div>
        <Divider />
        <Typography variant="body1" className={classes.subtitles}>
          Space Address
        </Typography>
        <div className={classes.location}>
          <div className={classes.address}>
            <Typography variant="body1">{address.address_1}</Typography>
            <Typography variant="body1">{address.address_2}</Typography>
            <Typography variant="body1">{address.city}</Typography>
            <Typography variant="body1">{address.postal_code}</Typography>
            <Typography variant="body1">{address.country}</Typography>
          </div>
          <div className={classes.distance}>
            <Typography variant="body1">TBD: distance</Typography>
          </div>
        </div>
        <Divider />
        <Typography variant="body1" className={classes.subtitles}>
          Phone Number
        </Typography>
        <Button color="primary" href={`tel:${phoneNumber}`}>
          {phoneNumber}
        </Button>
        <Divider />
        <Typography variant="body1" className={classes.subtitles}>
          WebSite
        </Typography>
        {website ? (
          <Link
            variant="body1"
            href={website}
            target="_blank"
            rel="noreferrer"
            className={classes.url}
          >
            {website}
          </Link>
        ) : (
          <Link
            variant="body1"
            href={yelpUrl}
            target="_blank"
            rel="noreferrer"
            className={classes.url}
          >
            Link to Yelp
          </Link>
        )}
        <Divider />
        <Typography variant="body1" className={classes.subtitles}>
          Hours Of Operation
        </Typography>
        {/* TODO:  fix placement of navigate next icon */}
        <Typography variant="body1" className={classes.mainInformation}>
          {hoursOfOperation ? 'Open Now' : 'Closed'}
        </Typography>
        {!hoursOfOperation && (
          <IconButton
            component="span"
            className={classes.nextButton}
            color="secondary"
            aria-label="hours of operation on yelp"
          >
            <a variant="body1" href={yelpUrl} target="_blank" rel="noreferrer">
              <NavigateNextIcon color="primary" />
            </a>
          </IconButton>
        )}
        <Divider />
        <Typography variant="body1" className={classes.subtitles}>
          Share
        </Typography>
        <Button
          color="primary"
          aria-label="visit space"
          component="span"
          className={classes.shareButton}
          onClick={handleShare}
        >
          <ShareIcon color="primary" style={{ padding: 0 }} />
        </Button>
        <Typography variant="body1" className={classes.mainInformation}>
          Share this Space with your network
        </Typography>
      </CardContent>
    </Card>
  );
};

SpaceDetailCard.propTypes = {
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  name: PropTypes.string.isRequired,
  phoneNumber: PropTypes.string.isRequired,
  filters: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired,
    }),
  ).isRequired,
  /* TODO: make conditional rendering for prop types?
   website: PropTypes.string,
  yelpUrl: PropTypes.string
  */
  category: PropTypes.string.isRequired,
  averageRating: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  imageUrl: PropTypes.string.isRequired,
  address: PropTypes.shape({}).isRequired,
  // TODO: add distance when geocoding
  // distance,
  hoursOfOperation: PropTypes.bool.isRequired,
  classes: PropTypes.shape({}).isRequired,
  overrideClasses: PropTypes.shape({}),
};

SpaceDetailCard.defaultProps = {
  overrideClasses: {},
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
};

export default withStyles(styles)(SpaceDetailCard);
