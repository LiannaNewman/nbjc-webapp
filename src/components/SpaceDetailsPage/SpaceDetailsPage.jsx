import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import StarIcon from '@material-ui/icons/Star';
import {
  Button,
  Typography,
  Checkbox,
  TextField,
  FormGroup,
  FormControlLabel,
} from '@material-ui/core';
import StarRating from '../StarRating';

const styles = {};

const SpaceDetailsPage = ({
  classes,
  name,
  category,
  averageRating,
}) => (
  <FormGroup>
    <h1> Rate and review this space.</h1>
    <h1> You can rate and review anonymously.</h1>
    <div
      aria-label="Name of Space"
      avatar={(
        <div color="secondary">
          <StarIcon color="secondary" fontSize="large" />
          <Typography
            variant="body2"
            color="secondary"
            className={classes.rating}
          >
            {averageRating}
          </Typography>
        </div>
    )}
      title={<Typography variant="h6">{name}</Typography>}
      subheader={<Typography variant="body2">{category}</Typography>}
    />

    <p>Rate this space.</p>
    <StarRating />

    <p>Share your thoughts</p>
    <TextField
      id="outlined-multiline-static"
      label="Write Review"
      multiline
      placeholder="Write Your Review Here"
      variant="outlined"
      required
    />

    <FormControlLabel
      control={<Checkbox name="publish" />}
      label="Publish my rating and review anonymously."
    />

    <Button color="secondary">Submit</Button>
  </FormGroup>
);

SpaceDetailsPage.propTypes = {
  name: PropTypes.string.isRequired,
  category: PropTypes.string.isRequired,
  averageRating: PropTypes.string.isRequired,
};

SpaceDetailsPage.defaultProps = {};

export default withStyles(styles)(SpaceDetailsPage);
