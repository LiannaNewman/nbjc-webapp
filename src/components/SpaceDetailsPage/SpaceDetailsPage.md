## SpaceDetailsPage

```jsx
const props = {
  name: 'La Colombe Coffee Roasters',
  category: 'Coffee',
  averageRating: '4.5',
};
<SpaceDetailsPage {...props} />;
```
