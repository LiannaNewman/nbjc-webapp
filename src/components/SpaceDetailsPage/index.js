import SpaceDetailsPage from './SpaceDetailsPage';

export default SpaceDetailsPage;
