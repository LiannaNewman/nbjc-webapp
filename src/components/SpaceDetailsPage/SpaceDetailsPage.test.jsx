import React from 'react';
import { render } from '@testing-library/react';

import SpaceDetailsPage from './SpaceDetailsPage';

test('renders SpaceDetailsPage', () => {
  const props = {
    name: 'La colombe coffee roasters',
    category: 'Coffee',
    averageRating: '4.5',
  };
  const { asFragment } = render(<SpaceDetailsPage {...props} />);
  expect(asFragment()).toMatchSnapshot();
});
