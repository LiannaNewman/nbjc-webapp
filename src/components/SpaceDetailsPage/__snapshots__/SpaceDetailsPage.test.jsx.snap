// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders SpaceDetailsPage 1`] = `
<DocumentFragment>
  <div
    class="MuiFormGroup-root"
  >
    <h1>
       Rate and review this space.
    </h1>
    <h1>
       You can rate and review anonymously.
    </h1>
    <div
      aria-label="Name of Space"
      avatar="[object Object]"
      subheader="[object Object]"
      title="[object Object]"
    />
    <p>
      Rate this space.
    </p>
    <button
      class="MuiButtonBase-root MuiIconButton-root StarRating-button-1 MuiIconButton-colorPrimary"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiIconButton-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
          />
        </svg>
      </span>
    </button>
    <button
      class="MuiButtonBase-root MuiIconButton-root StarRating-button-1 MuiIconButton-colorPrimary"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiIconButton-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
          />
        </svg>
      </span>
    </button>
    <button
      class="MuiButtonBase-root MuiIconButton-root StarRating-button-1 MuiIconButton-colorPrimary"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiIconButton-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
          />
        </svg>
      </span>
    </button>
    <button
      class="MuiButtonBase-root MuiIconButton-root StarRating-button-1 MuiIconButton-colorPrimary"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiIconButton-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
          />
        </svg>
      </span>
    </button>
    <button
      class="MuiButtonBase-root MuiIconButton-root StarRating-button-1 MuiIconButton-colorPrimary"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiIconButton-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M22 9.24l-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
          />
        </svg>
      </span>
    </button>
    <p>
      Share your thoughts
    </p>
    <div
      class="MuiFormControl-root MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
        data-shrink="false"
        for="outlined-multiline-static"
        id="outlined-multiline-static-label"
      >
        Write Review
        <span
          aria-hidden="true"
          class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
        >
           *
        </span>
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-multiline MuiOutlinedInput-multiline"
      >
        <textarea
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiOutlinedInput-inputMultiline"
          id="outlined-multiline-static"
          placeholder="Write Your Review Here"
          required=""
          rows="1"
          style="height: 0px; overflow: hidden;"
        />
        <textarea
          aria-hidden="true"
          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiOutlinedInput-inputMultiline"
          readonly=""
          style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); width: 100%;"
          tabindex="-1"
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-2 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-4"
          >
            <span>
              Write Review *
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <label
      class="MuiFormControlLabel-root"
    >
      <span
        aria-disabled="false"
        class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-6 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
      >
        <span
          class="MuiIconButton-label"
        >
          <input
            class="PrivateSwitchBase-input-9"
            data-indeterminate="false"
            name="publish"
            type="checkbox"
            value=""
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </span>
      <span
        class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
      >
        Publish my rating and review anonymously.
      </span>
    </label>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textSecondary"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Submit
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</DocumentFragment>
`;
